library/core/src/intrinsics/mir.rs

```diff
@@ -310,12 +310,43 @@ define!(
     fn UnwindCleanup(goto: BasicBlock)
 );
 
+// Return destination for `Call`
+define!("mir_return_to", fn ReturnTo(goto: BasicBlock));
+
 // Terminators
 define!("mir_return", fn Return() -> BasicBlock);
 define!("mir_goto", fn Goto(destination: BasicBlock) -> BasicBlock);
 define!("mir_unreachable", fn Unreachable() -> BasicBlock);
-define!("mir_drop", fn Drop<T, U>(place: T, goto: BasicBlock, unwind_action: U));
-define!("mir_call", fn Call<U>(call: (), goto: BasicBlock, unwind_action: U));
+define!("mir_drop",
+    /// Drop the contents of a place.
+    ///
+    /// The first argument must be a place.
+    ///
+    /// The second argument must be of the form `ReturnTo(bb)`, where `bb` is the basic block that
+    /// will be jumped to after the destructor returns.
+    ///
+    /// The third argument describes what happens on unwind. It can be one of:
+    /// - [`UnwindContinue`]
+    /// - [`UnwindUnreachable`]
+    /// - [`UnwindTerminate`]
+    /// - [`UnwindCleanup`]
+    fn Drop<T>(place: T, goto: (), unwind_action: ())
+);
+define!("mir_call",
+    /// Call a function.
+    ///
+    /// The first argument must be of the form `ret_val = fun(arg1, arg2, ...)`.
+    ///
+    /// The second argument must be of the form `ReturnTo(bb)`, where `bb` is the basic block that
+    /// will be jumped to after the function returns.
+    ///
+    /// The third argument describes what happens on unwind. It can be one of:
+    /// - [`UnwindContinue`]
+    /// - [`UnwindUnreachable`]
+    /// - [`UnwindTerminate`]
+    /// - [`UnwindCleanup`]
+    fn Call(call: (), goto: (), unwind_action: ())
+);
```

Comment:
aside:

I wonder if we should use ZSTs as the return type of these special arg fns like `ReturnTo` to make sure that we pass the right kind of args into the right intrinsics:

```rust
struct ReturnToArg;
// Return destination for `Call`
define!("mir_return_to", fn ReturnTo(goto: BasicBlock) -> ReturnToArg);

struct UnwindActionArg;
define!(
    "mir_unwind_unreachable",
    /// An unwind action that triggers undefined behaviour.
    fn UnwindUnreachable() -> UnwindActionArg
);
define!(
    "mir_unwind_terminate",
    /// An unwind action that terminates the execution.
    ///
    /// `UnwindTerminate` can also be used as a terminator.
    fn UnwindTerminate(reason: UnwindTerminateReason) -> UnwindActionArg
);
define!(
    "mir_unwind_cleanup",
    /// An unwind action that continues execution in a given basic block.
    fn UnwindCleanup(goto: BasicBlock) -> UnwindActionArg
);

define!("mir_call", fn Call(call: (), goto: ReturnToArg, unwind_action: UnwindActionArg));
```

Reply:
Not that it really matters because we're validating this in parsing later, 🤷

Reply:
The errors from parsing later are pretty bad, so I like this suggestion.